common/workunit/workunit.cpp

```diff
@@ -1804,6 +2304,480 @@ class CConstGraphProgressStatisticsIterator : public CInterfaceOf<IConstWUStatis
     bool valid;
 };
 
+//Extra an argument of the format abc[def[ghi...,]],... as (abc,def[...])
+static bool extractOption(const char * & finger, StringAttr & option, StringAttr & arg)
+{
+    const char * start = finger;
+    if (!*start)
+        return false;
+
+    const char * cur = start;
+    const char * bra = nullptr;
+    const char * end = nullptr;
+    unsigned braDepth = 0;
+    char next;
+    arg.clear();
+    for (;;)
+    {
+        next = *cur;
+        if (!next || ((next == ',') && (braDepth == 0)))
+            break;
+
+        switch (next)
+        {
+        case '[':
+            if (braDepth == 0)
+            {
+                if (bra)
+                    throw makeStringExceptionV(0, "Multiple [ in filter : %s", bra);
+                bra = cur;
+            }
+            braDepth++;
+            break;
+        case ']':
+            if (braDepth > 0)
+            {
+                if (--braDepth == 0)
+                {
+                    end = cur;
+                    arg.set(bra+1, cur - (bra+1));
+                }
+            }
+            break;
+        }
+
+        cur++;
+    }
+
+    if (braDepth != 0)
+        throw makeStringExceptionV(0, "Mismatched ] in filter : %s", start);
+
+    if (bra)
+    {
+        if (cur != end+1)
+            throw makeStringExceptionV(0, "Text follows closing bracket: %s", end);
+        option.set(start, bra-start);
+    }
+    else
+        option.set(start, cur-start);
+
+    if (next)
+        finger = cur+1;
+    else
+        finger = cur;
+    return true;
+}
+
+enum { FOscope, FOstype, FOid, FOdepth, FOsource, FOwhere, FOmatched, FOnested, FOinclude, FOproperties, FOstatistic, FOattribute, FOhint, FOmeasure, FOversion }; // Enum to clarify the swicth statement below
+static constexpr const char * filterOptions[] = { "scope", "stype", "id", "depth", "source", "where", "matched", "nested", "include", "prop", "stat", "attr", "hint", "measure", "version", nullptr };
+static constexpr EnumMapping sourceMappings[] = { { SSFsearchGlobalStats, "global" }, { SSFsearchGraphStats, "stats" }, { SSFsearchGraph, "graph" }, { SSFsearchExceptions, "exception" }, { (int)SSFsearchAll, "all" }, { 0, nullptr } };
+static constexpr EnumMapping propertyMappings[] = { { PTnone, "none" }, { PTstatistics, "stat" }, { PTattributes, "attr" }, { PThints, "hint" }, { PTscope, "scope" }, { PTall, "all" }, { 0, nullptr } };
+WuScopeFilter::WuScopeFilter(const char * filter)
+{
+    addFilter(filter);
+    finishedFilter();
+}
+
+void WuScopeFilter::addFilter(const char * filter)
+{
+    if (!filter)
+        return;
+
+    StringAttr option;
+    StringAttr arg;
+    while (extractOption(filter, option, arg))
+    {
+        switch (matchString(option, filterOptions))
+        {
+        case FOscope:
+            addScope(arg);
```

Comment:
Scopes aren't validated here  -wouldn't it be safer to?

Reply:
How would you suggest validating them, and what advantage would it give?

Reply:
Check it is a valid sequence of id:id:id?

Reply:
Instead of modifying this PR, and requiring retesting I have added a requirement to https://track.hpccsystems.com/browse/HPCC-18262 to validate the string correctly.